File: razer-sdk/src/main/java/com/bonepl/razersdk/color/TransitionColor.java
package com.bonepl.razersdk.color;

/**
 * Dynamic Color that continuously provides
 * Color being the result of transition between provided fromColor and toColor.
 * Once the transition finishes - toColor is returned
 */
public class TransitionColor implements Color {
    private static final int DEFAULT_STEPS = 20;
    private StaticColor from;
    private StaticColor to;
    private int steps;
    private int currentStep;

    public TransitionColor(StaticColor from, StaticColor to) {
        this(from, to, DEFAULT_STEPS);
    }

    public TransitionColor(StaticColor from, StaticColor to, int steps) {
        this.from = from;
        this.to = to;
        if (steps < 2) {
            throw new IllegalArgumentException("TransitionColor needs at least 2 steps for transition, provided steps: " + steps);
        }
        this.steps = steps;
    }

    public StaticColor getColorAtPercent(int percent) {
        return new StaticColor((int) (from.red() - (from.red() - to.red()) * percent * 0.01),
                (int) (from.green() - (from.green() - to.green()) * percent * 0.01),
                (int) (from.blue() - (from.blue() - to.blue()) * percent * 0.01));
    }

    private int getRedStep() {
        return (from.red() - to.red()) / (steps - 1);
    }

    private int getGreenStep() {
        return (from.green() - to.green()) / (steps - 1);
    }

    private int getBlueStep() {
        return (from.blue() - to.blue()) / (steps - 1);
    }

    public boolean transitionFinished() {
        return currentStep == steps;
    }

    public void resetTransition() {
        currentStep = 0;
    }

    public void setFrom(StaticColor from) {
        this.from = from;
    }

    public void setTo(StaticColor to) {
        this.to = to;
    }

    public void setSteps(int newSteps) {
        if (newSteps != steps) {
            final double stepsRatio = (double) currentStep / steps;
            currentStep = (int) (stepsRatio * newSteps);
            steps = newSteps;
        }
    }

    @Override
    public StaticColor getColor() {
        if (transitionFinished()) {
            return to;
        }

        if(currentStep + 1 == steps){
            currentStep += 1;
            return to;
        }

        final StaticColor color = new StaticColor(from.red() - getRedStep() * currentStep,
                from.green() - getGreenStep() * currentStep,
                from.blue() - getBlueStep() * currentStep);
        currentStep += 1;
        return color;
    }
}

File: src/razer-sdk/color/TransitionColor.ts
import {Color} from "./Color.js";
import {StaticColor} from "./StaticColor.js";

const DEFAULT_STEPS = 20;

export class TransitionColor implements Color {
    private from: StaticColor;
    private to: StaticColor;
    private steps: number;
    private currentStep: number = 0;

    constructor(from: StaticColor, to: StaticColor, steps: number = DEFAULT_STEPS) {
        this.from = from;
        this.to = to;
        if (steps < 2) {
            throw new Error(
                "TransitionColor needs at least 2 steps for transition, provided steps: " + steps
            );
        }
        this.steps = steps;
    }

    getColorAtPercent(percent: number): StaticColor {
        return new StaticColor(
            Math.trunc(this.from.red - (this.from.red - this.to.red) * percent * 0.01),
            Math.trunc(this.from.green - (this.from.green - this.to.green) * percent * 0.01),
            Math.trunc(this.from.blue - (this.from.blue - this.to.blue) * percent * 0.01)
        );
    }

    private getRedStep(): number {
        return Math.round((this.from.red - this.to.red) / (this.steps - 1));
    }

    private getGreenStep(): number {
        return Math.round((this.from.green - this.to.green) / (this.steps - 1));
    }

    private getBlueStep(): number {
        return Math.round((this.from.blue - this.to.blue) / (this.steps - 1));
    }

    transitionFinished(): boolean {
        return this.currentStep === this.steps;
    }

    resetTransition(): void {
        this.currentStep = 0;
    }

    setFrom(from: StaticColor): void {
        this.from = from;
    }

    setTo(to: StaticColor): void {
        this.to = to;
    }

    setSteps(newSteps: number): void {
        if (newSteps !== this.steps) {
            const stepsRatio = this.currentStep / this.steps;
            this.currentStep = Math.trunc(stepsRatio * newSteps);
            this.steps = newSteps;
        }
    }

    getColor(): StaticColor {
        if (this.transitionFinished()) {
            return this.to;
        }

        if (this.currentStep + 1 === this.steps) {
            this.currentStep += 1;
            return this.to;
        }

        const color = new StaticColor(
            this.from.red - this.getRedStep() * this.currentStep,
            this.from.green - this.getGreenStep() * this.currentStep,
            this.from.blue - this.getBlueStep() * this.currentStep
        );
        this.currentStep += 1;
        return color;
    }
}
